File: app/schemas/telemetry.py
from datetime import datetime, timezone
from enum import Enum
from typing import Any
from uuid import UUID, uuid4

from pydantic import BaseModel, Field, field_serializer


class OperationName(str, Enum):
    LLM_PROMPT = "llm_prompt"
    EXECUTE_TOOL = "execute_tool"
    STATE_TRANSITION = "state_transition"


class ToolCategory(str, Enum):
    DATABASE_WRITE = "database_write"
    FILE_EGRESS = "file_egress"
    SYSTEM_EXEC = "system_exec"
    READ = "read"


class AgentMeta(BaseModel):
    agent_id: str
    framework: str = "langgraph"
    identity_urn: str
    delegation_chain: list[str] = Field(default_factory=list)


class ToolMeta(BaseModel):
    name: str
    category: ToolCategory
    call_id: str
    parameters: dict[str, Any] = Field(default_factory=dict)


class TelemetryEvent(BaseModel):
    event_id: UUID = Field(default_factory=uuid4)
    trace_id: UUID
    parent_span_id: UUID | None = None
    timestamp: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    agent: AgentMeta
    operation_name: OperationName
    tool: ToolMeta | None = None
    payload_content: str
    sanitized_payload: str | None = None
    detections: dict[str, Any] | None = None

    @field_serializer("timestamp")
    def serialize_timestamp(self, dt: datetime, _info: Any) -> str:
        return dt.isoformat()

File: frontend/src/types/telemetry.ts
export type OperationName = 'llm_prompt' | 'execute_tool' | 'state_transition';
export type ToolCategory = 'database_write' | 'file_egress' | 'system_exec' | 'read';
export type SeverityLevel = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
export type ContainmentStatus = 'CONTAINED' | 'PARTIALLY_CONTAINED' | 'FAILED' | 'NO_ACTION';
export type ContainmentTier = 'tier_1_soft' | 'tier_2_medium' | 'tier_3_hard';

export interface AgentMeta {
  agent_id: string;
  framework: string;
  identity_urn: string;
  delegation_chain: string[];
}

export interface ToolMeta {
  name: string;
  category: ToolCategory;
  call_id: string;
  parameters: Record<string, any>;
}

export interface FastPathResult {
  matched: boolean;
  rule_id?: string;
  rule_name?: string;
  latency_ms: number;
}

export interface PolicyResult {
  allowed: boolean;
  identity_urn: string;
  tool_category?: ToolCategory;
  reason?: string;
}

export interface JudgeVerdict {
  is_anomalous: boolean;
  confidence: number;
  divergence_score: number;
  rationale: string;
}

export interface EventDetections {
  fastpath?: FastPathResult;
  policy?: PolicyResult;
  judge?: JudgeVerdict;
}

export interface TelemetryEvent {
  event_id: string;
  trace_id: string;
  parent_span_id?: string | null;
  timestamp: string;
  agent: AgentMeta;
  operation_name: OperationName;
  tool?: ToolMeta | null;
  payload_content: string;
  sanitized_payload?: string | null;
  detections?: EventDetections;
}

export interface ContainmentActionResult {
  tier: ContainmentTier;
  action_name: string;
  success: boolean;
  details: string;
}

export interface ContainmentResult {
  incident_id: string;
  status: ContainmentStatus;
  executed_tiers: ContainmentTier[];
  action_results: ContainmentActionResult[];
}

export interface IncidentRecord {
  incident_id: string;
  trace_id: string;
  severity: SeverityLevel;
  status: ContainmentStatus;
  agent_id: string;
  identity_urn: string;
  matched_techniques: string[];
  rationale: string;
  containment_result?: ContainmentResult;
  created_at: string;
  updated_at?: string;
}

export interface SystemHealth {
  status: string;
  project: string;
  mongodb?: string;
  judge_mode?: string;
}
